### PR TITLE
Incoming stamina damage while in stamcrit has diminishing returns applied

### PR DESCRIPTION
**Pull request by MrMelbert**
## About The Pull Request

After being put in stamcrit, future incoming stamina damage has "diminishing returns" applied*. 

The formula looks like `ceil(sqrt(amount of stamina damage) / 2) - times you have taken stamina damage in stamcrit`. 

This means eventually stamina based damage will do less than zero damage, and thus, not contribute to keeping stamcrit active.

Very, very low amounts of stamina damage (such as from chems) contribute to DR 1/20th the amount. 

*_Note, this is not real diminishing returns because making it real diminishing returns would be pointless, you are capped to 120 stamina damage so 99.99% of the time you take stam damage while in stam crit you're already capped. This is just faking the effect._

In its current stat this means that a stun baton will stop being able to keep someone in stamcrit after the 5th hit, and a stock disabler will stop being able to keep someone in stamcrit after the 4th hit. 

## Why It's Good For The Game

Mostly just an experiment. I don't imagine it will shake up much about the baton situation in its current state. 

There's also no grace period after getting UP - so like you can just be stamcritted right after anyways. Maybe there should be one? Food for thought.

And yes of course you can just space out your hits, you're not clever for thinking about that

## Changelog

:cl: Melbert
balance: Taking stamina damage in stamcrit has diminishing returns associated, meaning you cannot be infinitely stamcrit.  
/:cl:

### CODE REVIEW

**Comment by MrMelbert** [comment]
I should mention it's extremely unlikely for this to affect sec gameplay in any major way, this is mostly targeted at sources of infinite stamina damage we have - See #83444

**Comment by carlarctg** [comment]
nerfs stun batons, close

**Comment by DaydreamIQ** [comment]
Honestly with how oppressive telebatons can be in a 1v1 this is a welcome change

**Comment by Higgin** [comment]
good!

**Comment by moocowswag** [comment]
worried there might be some unintended side effect with https://github.com/tgstation/tgstation/pull/46323 since it forcefully increases the amount of stam you have after you enter stamcrit, wouldnt this pr interact with it?

**Comment by MrMelbert** [comment]
> worried there might be some unintended side effect with #46323 since it forcefully increases the amount of stam you have after you enter stamcrit, wouldnt this pr interact with it?

No that's specifically counted for, look at the code

**Comment by YuliceTheDruid** [comment]
License to murder for heads of staff 😂

File: code/__DEFINES/mobs.dm [hunk 1]
```diff
@@ -161,7 +161,12 @@
 #define HUMAN_MAX_OXYLOSS 3
 #define HUMAN_CRIT_MAX_OXYLOSS (SSMOBS_DT/3)
 
-#define STAMINA_REGEN_BLOCK_TIME (10 SECONDS)
+/// How long it takes to return to 0 stam
+#ifdef UNIT_TESTS
```
Comment (Mothblocks) [inline code/__DEFINES/mobs.dm:165]: why are we doing this?

Comment (Mothblocks) [reply code/__DEFINES/mobs.dm:165]: ugh. because we don't have time stepping. leave a comment

File: code/modules/unit_tests/combat_stamina.dm [hunk 2]
```diff
@@ -0,0 +1,20 @@
+/// Tests 100 stamina damage = stamcrit
+/datum/unit_test/stamcrit
```
Comment (Mothblocks) [inline code/modules/unit_tests/combat_stamina.dm:2]: any tests that intentionally sleep ought to be put into the slow tests categories

File: code/modules/mob/living/damage_procs.dm [hunk 3]
```diff
@@ -451,26 +451,42 @@
...
+	return delta
+
+/// The mob has received stamina damage
+///
+/// - current_level: The mob's current stamina damage amount (to save unnecessary getStaminaLoss() calls)
+/// - amount_actual: The amount of stamina damage received, in actuality
+/// For example, if you are taking 50 stamina damage but are at 90, you would actually only receive 30 stamina damage (due to the cap)
+/// - amount: The amount of stamina damage received, raw
+/mob/living/proc/received_stamina_damage(current_level, amount_actual, amount)
+	addtimer(CALLBACK(src, PROC_REF(setStaminaLoss), 0, TRUE, TRUE), STAMINA_REGEN_TIME, TIMER_UNIQUE|TIMER_OVERRIDE)
```
Comment (Mothblocks) [inline code/modules/mob/living/damage_procs.dm:489]: honestly i think this `STAMINA_REGEN_TIME` should be a global/ss var somewhere and the unit test changes it directly. we really should not have such broad surface area for this test

**Comment by ATH1909** [comment]
The big impact I see from this is on borgs, who lack cuffs and have nonlethal combat (from flashes, haloperidol, and the stun arm) as a significant part of their kits. The flash burnout mechanic did already previously limit borg chain stunning somewhat, but this PR creates a firmer one.

Being unable to chainstun with the gamer arm until your charge runs out will also have a significant impact on malf AIs going for cyborg conversion machine strats, I think.

**Comment by MrMelbert** [comment]
Polling over this now to see if anyone actually noticed or cared

image

**Comment by MrMelbert** [comment]
image

🤷‍♂️

**Pull request merged**